### PR TITLE
36113: Support CSRF

### PR DESCRIPTION
**Pull request by labkey-nicka**
- allows support for targeting v19.1+ LabKey instances
- request CSRF from login-whoAmI.api action
- enable cookies on UserAgent to maintain session
- coalesce handling of common parameters, authentication creation,
building urls, and establishing server context

### CODE REVIEW

File: src/lib/LabKey/Query.pm [hunk 1]
```diff
@@ -140,78 +141,37 @@ NOTE:
...
-	my $lk_config;
-	my $netrc_file = $args{-netrcFile} || $ENV{LABKEY_NETRC};
-	if(!$args{'-loginAsGuest'}){
-		$lk_config = _readrc( $args{-machine}, $netrc_file );
+	my $data = {
+        'schemaName'      => $args{'-schemaName'},
+        'query.queryName' => $args{'-queryName'}
+    };
+
+	if ($args{'-requiredVersion'}) {
```
Comment (labkey-kevink) [inline src/lib/LabKey/Query.pm]: Remove the note in docs above selectRows that requiredVersion now defaults to 8.3 to match other LabKey api libraries.

File: src/lib/LabKey/Query.pm [hunk 2]
```diff
@@ -140,78 +141,37 @@ NOTE:
...
+	foreach (@{$args{-parameters}}) {
+        $$data{'query.param.' . @{$_}[0]} = @{$_}[1];
 	}
 	
-	foreach ('viewName', 'offset', 'sort', 'maxRows', 'columns', 'containerFilterName'){
-		if ( $args{'-'.$_} ) {
-			$params{"query.".$_} = $args{'-'.$_};
+	foreach ('viewName', 'offset', 'sort', 'maxRows', 'columns', 'containerFilterName') {
+		if ($args{'-'.$_}) {
+            $$data{'query.'.$_} = $args{'-'.$_};
```
Comment (labkey-kevink) [inline src/lib/LabKey/Query.pm]: Include spaces around the '.' operator.

```suggestion
            $$data{'query.' . $_} = $args{'-' . $_};
```

File: src/lib/LabKey/Query.pm [hunk 3]
```diff
@@ -478,58 +359,27 @@ NOTE:
...
 	my $data = {
 		"schemaName" => $args{'-schemaName'},
-		"sql" => $args{'-sql'},			
+		"sql"        => $args{'-sql'}
 	};
 	
-	foreach ('offset', 'sort', 'maxRows', 'containerFilterName'){
-		if ( $args{'-'.$_} ) {
+	foreach ('offset', 'sort', 'maxRows', 'containerFilterName') {
+		if ($args{'-'.$_}) {
```
Comment (labkey-kevink) [inline src/lib/LabKey/Query.pm]:
```suggestion
		if ($args{'-' . $_}) {
```

**Review by labkey-kevink — APPROVED** [review]
Looks good

File: src/lib/LabKey/Query.pm (hunk 1 above)
Comment (labkey-nicka) [reply src/lib/LabKey/Query.pm]: Good call, meant to change this back to 9.1 default. Can later be adjusted to 8.3 if it is decided that a major release version is warranted.

**Pull request merged**